opengever/maintenance/scripts/use_canonical_document_ids.py

```diff
@@ -0,0 +1,135 @@
+from csv import DictWriter
+from datetime import datetime
+from opengever.document.behaviors.name_from_title import IDocumentNameFromTitle
+from opengever.maintenance.debughelpers import setup_app
+from opengever.maintenance.debughelpers import setup_option_parser
+from opengever.maintenance.debughelpers import setup_plone
+from opengever.maintenance.utils import elevated_privileges
+from plone import api
+from urlparse import urljoin
+import os
+import transaction
+import urllib
+
+
+SEPARATOR = '-' * 78
+
+
+REPORT_PATH = os.path.abspath(
+    os.path.join(__file__,
+                 '..', '..', '..', '..',
+                 'renamed_documents.csv'))
+
+
+def dev_break_document_ids(options):
+    """DEV-method, use to break the document ids.
+
+    """
+    catalog = api.portal.get_tool('portal_catalog')
+    brains = catalog.unrestrictedSearchResults(
+        portal_type='opengever.document.document')
+
+    for number, brain in enumerate(brains):
+        old_name = brain.id
+        new_name = "with spaces {}".format(number)
+        obj = brain.getObject()
+        api.content.rename(obj=obj, new_id=new_name)
+        print "broke {} to {} ({})".format(old_name, new_name,
+                                           '/'.join(obj.getPhysicalPath()))
+
+    if not options.dry_run:
+        transaction.commit()
+    print "done breaking."
+
+
+def use_canonical_document_id(options):
+    """Rename documents that do not use the current naming-schema.
+
+    Also write a short report of the renamed documents to a csv file.
+
+    """
+    catalog = api.portal.get_tool('portal_catalog')
+    brains = catalog.unrestrictedSearchResults(
+        portal_type='opengever.document.document')
+
+    to_rename = []
+
+    for brain in brains:
+        name = brain.id
+        if name.startswith('template-') or name.startswith('document-'):
+            if options.verbose:
+                print "skipped: {}".format(brain.getPath())
+            continue
+
+        to_rename.append(brain.getObject())
+
+    if to_rename:
+        print SEPARATOR
+
+    def build_url(fragment):
+        fragment = urllib.quote(fragment)
+        if not options.domain:
+            return fragment
+
+        # we assume a virtualhost and strip the plone site id
+        assert fragment.startswith("/"), "path should start with /"
+        visible_fragment = "/".join(fragment.split('/')[2:])
+        base = "https://{}/".format(options.domain)
+        return urljoin(base, visible_fragment)
+
+    with open(REPORT_PATH, 'w+') as csvfile:
+        writer = DictWriter(csvfile, fieldnames=["old_url", "new_url"])
+        writer.writeheader()
+
+        # elevated privileges are necessary to be able to modify stuff in
+        # closed dossiers.
+        with elevated_privileges():
+            for obj in to_rename:
+                old_name = obj.getId()
+                old_url = '/'.join(obj.getPhysicalPath())
+                name_chooser = IDocumentNameFromTitle(obj)
```

Comment:
I would instead use the `INameChooser` here, which will automatically [delegate to a `INameFromTitle` behavior if present](https://github.com/plone/plone.app.content/blob/1782199cf70eb54c6d17d1daedd694496432ba4d/plone/app/content/namechooser.py#L34).

So:
```python
new_name = INameChooser(aq_parent(obj)).chooseName(None, obj)
```

(From [`plone.dexterity.utils.addContentToContainer`](https://github.com/plone/plone.dexterity/blob/e59ed59b6a6914ec269bf028ab315f93f39f1220/plone/dexterity/utils.py#L179))

The conversion to a bytestring can be omitted, the `INameChooser` will already take care of that.